app/controllers/users/sessions_controller.rb

```diff
@@ -44,22 +50,29 @@ def require_no_authentication
   end
 
   def user_params
-    params.require(:user).permit(:email, :password, :remember_me, :otp_attempt)
+    params.require(:user).permit(:email, :password, :remember_me, :otp_attempt, :mobile_number)
   end
 
   def valid_otp_attempt?(user)
     user.validate_and_consume_otp!(user_params[:otp_attempt])
   end
 
-  def authenticate_with_otp_two_factor
+  def handle_otp_flow
```

Comment:
I think this method would benefit from a description (or link to the description) of the intended handling.

Reply:
I agree, but am wondering if we can factor it differently such that it explains itself a lot better. We're pushing a lot through a single POST to a `create` method and then `if`ing it up...

Reply:
Here's an idea:

Get rid of `handle_otp_flow` altogether, and stick to (and admittedly push to the limits of) the approach that devise is expecting, which is to prepend an action to create for each thing we want to check before authenticating:

```ruby
  prepend_before_action :ask_for_mobile_number, only: :create, if: -> { user_missing_mobile_number? }
  prepend_before_action :save_new_mobile_number, only: :create, if: -> { user_provided_new_mobile_number? }
  prepend_before_action :prompt_for_otp, only: :create, if: -> { user_has_mobile_number? }
  prepend_before_action :validate_otp, only: :create
```

Reply:
Oh, that's nice, I see what you're driving at. I agree with you that this is perhaps pushing `prepend_before_action` a little far, though! How about something that takes symbols like that and makes our little state machine clear (and hopefully also debuggable!) in one method?

```ruby
  def handle_otp_flow
    case phase
    when :valid_password_needs_mobile_number then prompt_for "mobile_number"
    when :valid_password_has_mobile_number then send_and_prompt_for_otp
    when :updating_mobile_number
      user.update!(mobile_number: user_params[:mobile_number])
      send_and_prompt_for_otp
    when :validating_otp then authenticate_user_with_otp_two_factor
    end
  end
```

Reply:
I think that's definitely an improvement! Better names needed now I think. Maybe `phase` can become something like `mfa_state` and `handle_otp_flow` can become `mfa_authenticate`.

As for the state names, aren't they always going to have valid_password? Do we need to state that twice?